crates/key-derivation/src/operator.rs

```diff
@@ -0,0 +1,268 @@
+//! Key derivation for Strata bridge operators
+//!
+//! Bridge operators guarantee the security assumptions of the Strata BitVM-based
+//! bridge by enforcing that all peg-ins and peg-outs are valid.
+//!
+//! Operators are responsible for their own keys and master [`Xpriv`] is not
+//! shared between operators. Hence, this crate has a BYOK (Bring Your Own Key) design.
+//!
+//! They use a set of keys to sign messages and bitcoin transactions.
+//! The keys are derived from a master [`Xpriv`]
+//! using a [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+//! HD derivation path.
+//!
+//! The derivation paths are:
+//!
+//! - `m/56'/20'/100` for the message signing key
+//! - `m/56'/20'/101` for the wallet transaction signing key
+
+use bitcoin::bip32::{ChildNumber, DerivationPath, Xpriv, Xpub};
+use secp256k1::SECP256K1;
+#[cfg(feature = "zeroize")]
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+use crate::error::KeyError;
+
+/// The Strata base index for operator keys.
+const BASE_IDX: u32 = 56;
+
+/// The operator index for operator keys.
+const OPERATOR_IDX: u32 = 20;
+
+/// The message index for the operator message key.
+const MESSAGE_IDX: u32 = 100;
+
+/// The wallet index for the operator wallet key.
+const WALLET_IDX: u32 = 101;
+
+/// The operator's message signing and wallet transaction signing keys.
+#[derive(Debug, Clone)]
+pub struct OperatorKeys {
+    /// Operator's master [`Xpriv`].
+    master: Xpriv,
+    /// Operator's message signing [`Xpriv`].
+    signing: Xpriv,
+    /// Operator's wallet transaction signing [`Xpriv`].
+    wallet: Xpriv,
+}
+
+impl OperatorKeys {
+    /// Creates a new [`OperatorKeys`] from a master [`Xpriv`].
+    pub fn new(master: &Xpriv) -> Result<Self, KeyError> {
+        let message_path = DerivationPath::master().extend([
+            ChildNumber::from_hardened_idx(BASE_IDX).unwrap(),
+            ChildNumber::from_hardened_idx(OPERATOR_IDX).unwrap(),
+            ChildNumber::from_normal_idx(MESSAGE_IDX).unwrap(),
+        ]);
+
+        let wallet_path = DerivationPath::master().extend([
+            ChildNumber::from_hardened_idx(BASE_IDX).unwrap(),
+            ChildNumber::from_hardened_idx(OPERATOR_IDX).unwrap(),
+            ChildNumber::from_normal_idx(WALLET_IDX).unwrap(),
+        ]);
+
+        let message_xpriv = master.derive_priv(SECP256K1, &message_path)?;
+        let wallet_xpriv = master.derive_priv(SECP256K1, &wallet_path)?;
+
+        Ok(Self {
+            master: *master,
+            signing: message_xpriv,
+            wallet: wallet_xpriv,
+        })
+    }
+
+    /// Operator's master [`Xpriv`].
+    pub fn master_xpriv(&self) -> Xpriv {
+        self.master
+    }
+
+    /// Operator's wallet transaction signing [`Xpriv`].
+    pub fn wallet_xpriv(&self) -> Xpriv {
+        self.wallet
+    }
+
+    /// Operator's message signing [`Xpriv`].
+    pub fn message_xpriv(&self) -> Xpriv {
+        self.signing
+    }
```

Comment:
Exposing these by value seems like it introduces risk, as these values are not zeroized on drop. Does exposing them by reference play nicely with intended use cases?

Reply:
The problem is that all of the crates depend on owning an `Xpriv` to sign.
See for example https://github.com/alpenlabs/strata/blob/b5f84a04fe1f45503cb70da076e93abe3fc31554/crates/bridge-sig-manager/src/manager.rs#L41-L42 and https://github.com/alpenlabs/strata/blob/b5f84a04fe1f45503cb70da076e93abe3fc31554/crates/bridge-exec/src/handler.rs#L39-L40 (`Keypair` is derived from an `Xpriv`) or https://github.com/alpenlabs/strata/blob/b5f84a04fe1f45503cb70da076e93abe3fc31554/bin/bridge-client/src/modes/operator/bootstrap.rs#L70-L75

Reply:
Good catch done in 13ab5a9ef6c9f83677eb9e835538127e9e8a30ee